runtime/cudaq/utils/details/tensor_impl.h

```diff
@@ -0,0 +1,183 @@
+  /// @return A `const` reference to the element at the specified indices
+  virtual const scalar_type &at(const std::vector<size_t> &indices) const = 0;
+
+  /// @brief Copy data into the tensor
+  /// @param data Pointer to the source data
+  /// @param shape The shape of the source data
+  virtual void copy(const scalar_type *data,
+                    const std::vector<std::size_t> &shape) = 0;
+
+  /// @brief Take ownership of the given data
+  /// @param data Pointer to the source data
+  /// @param shape The shape of the source data
+  virtual void take(std::unique_ptr<scalar_type[]> &data,
+                    const std::vector<std::size_t> &shape) = 0;
+  virtual void take(std::unique_ptr<scalar_type[]> &&data,
+                    const std::vector<std::size_t> &shape) = 0;
+
+  /// @brief Borrow the given data without taking ownership
+  /// @param data Pointer to the source data
+  /// @param shape The shape of the source data
+  virtual void borrow(const scalar_type *data,
+                      const std::vector<std::size_t> &shape) = 0;
+
+  /// @brief Get a pointer to the raw data of the tensor.
+  /// This method provides direct access to the underlying data storage of the
+  /// tensor. It returns a pointer to the first element of the data array.
+  ///
+  /// @return scalar_type* A pointer to the mutable data of the tensor.
+  /// @note Care should be taken when directly manipulating the raw data to
+  /// avoid invalidating the internal state of the tensor or violating its
+  /// `invariants`.
+  virtual scalar_type *data() = 0;
+
+  /// @brief Get a constant pointer to the raw data of the tensor.
+  /// This method provides read-only access to the underlying data storage of
+  /// the tensor. It returns a constant pointer to the first element of the data
+  /// array.
+  ///
+  /// @return constant scalar_type * A constant pointer to the immutable data of
+  /// the tensor.
+  /// @note This constant version ensures that the data of the tensor cannot be
+  /// modified through the returned pointer, preserving `const` correctness.
+  virtual const scalar_type *data() const = 0;
+
+  virtual void dump() const = 0;
+
+  virtual ~tensor_impl() = default;
+
+  // Operator friends.
+  template <typename T>
+  friend tensor_impl<T> *operator*(const tensor_impl<T> &,
+                                   const tensor_impl<T> &);
+  template <typename T>
+  friend tensor_impl<T> *operator+(const tensor_impl<T> &,
+                                   const tensor_impl<T> &);
+
+  // Double-dispatch hooks. We use double dispatch to ensure that both arguments
+  // are in fact the same derived class of `tensor_impl`.
+  virtual tensor_impl<Scalar> *
+  dd_multiply(const tensor_impl<Scalar> &left) const = 0;
+  virtual tensor_impl<Scalar> *
+  dd_add(const tensor_impl<Scalar> &left) const = 0;
+
+  // Terminal implementation of operators.
+  virtual tensor_impl<Scalar> *multiply(const xtensor<Scalar> &right) const = 0;
```

Comment:
is the use of xtensor intended here?

Reply:
Yes. This is using the old-school double dispatch pattern to determine the two input tensors have the same implementation. This is needed so that our matrix multiply code can efficiently (and correctly) access both of the input matrices to compute the product.

I'm thinking this implementation hierarchy ought to get rid of the virtual methods and be redone using the curiously recurring template pattern. However, I left it as it was to this point.